docs/source/extension/extension_points.rst

```diff
@@ -1135,3 +1135,181 @@ Occasionally, LSP extensions include a CodeMirror extension to modify the code e
+        return {
+          ...model,
+          content: []
+        };
+      }
+    }
+
+    // ...
+  }
+
+  const customContentProviderPlugin: JupyterFrontEndPlugin<void> = {
+    id: 'my-extension:custom-content-provider',
+    autoStart: true,
+    activate: (app: JupyterFrontEnd) => {
+      const drive = (app.serviceManager.contents as ContentsManager).defaultDrive;
+      const registry = drive?.contentProviderRegistry;
+      if (!registry) {
+        // If content provider is a non-essential feature and support for JupyterLab <4.4 is desired:
+        console.error('Cannot initialize content provider: no content provider registry.');
+        return;
+      }
+      const customContentProvider = new CustomContentProvider({
+        // These options are only required if extending the `RestContentProvider`.
+        apiEndpoint: '/api/contents',
+        serverSettings: app.serviceManager.serverSettings,
+      });
+      registry.register('my-custom-provider', customContentProvider);
+    }
+  };
+
+and then create and register a widget factory which will understand how to make use of your custom content provider:
+
+.. code:: typescript
+
+  class ExampleWidgetFactory extends ABCWidgetFactory<ExampleDocWidget, ExampleDocModel> {
+    protected createNewWidget(
+      context: DocumentRegistry.IContext<ExampleDocModel>
+    ): ExampleDocWidget {
+
+      return new ExampleDocWidget({
+        context,
+        content: new ExamplePanel(context)
+      });
+    }
+  }
+
+  const widgetFactoryPlugin: JupyterFrontEndPlugin<void> = {
+    id: 'my-extension:custom-widget-factory',
+    autoStart: true,
+    activate: (app: JupyterFrontEnd) => {
+
+      const widgetFactory = new ExampleWidgetFactory({
+        name: FACTORY,
+        modelName: 'example-model',
+        fileTypes: ['example'],
+        defaultFor: ['example'],
+        // Instructs the document registry to use the custom provider
+        // for context of widgets created with `ExampleWidgetFactory`.
+        contentProviderId: 'my-custom-provider'
+      });
+      app.docRegistry.addWidgetFactory(widgetFactory);
+    }
+  };
+
+
+Where ``ExampleDocModel`` can now expect the ``CustomContentsModel`` rather than ``Contents.IModel``:
+
+.. code:: typescript
+
+  class ExampleDocModel implements DocumentRegistry.IModel {
+    // ...
+
+    fromJSON(chunks: IMyContentChunk[]): void {
+      this.sharedModel.transact(() => {
+        let i = 0;
+        for (const chunk of chunks) {
+          const chunk = fetch(chunk.url);
+          this.sharedModel.set(`chunk-${i}`, chunk);
+          i += 1;
+        }
+      });
+    }
+
+    fromString(data: string): void {
+      const chunks = JSON.parse(data) as IMyContentChunk[];
+      return this.fromJSON(chunks);
+    }
+  }
+
+
+For a complete example of a widget factory (although not using a content provider), see the `documents extension example <https://github.com/jupyterlab/extension-examples/tree/main/documents>`__.
```

Comment:
Linking to https://github.com/jupyterlab/extension-examples/issues/215 as potentially related.

At some point it will be great to have an example about contents in the `extension-examples` repository. Not sure yet if there should be two distinct examples for both drives and content providers, or only one so it's easier to explain the difference between the two APIs and how they are supposed to be used together.

Reply:
> if there should be two distinct examples for both drives and content providers, or only one so it's easier to explain the difference between the two APIs and how they are supposed to be used together

I think there is a big advantage of having drives and content providers in a single example, but on the other hand we also need a widget factory so that gets a pretty big. Maybe in addition to one such a large example there should be a small `IDrive` example.

Reply:
Sounds good, we can track that and discuss more in the `extension-examples` repo.